sql/parser/normalize.go

```diff
@@ -0,0 +1,359 @@
+				break
+			default:
+				return v, n
+			}
+			// The left side is const and the right side is a binary expression or a
+			// variable. Flip the comparison op so that the right side is const and
+			// the left side is a binary expression or variable.
+			n.Operator = invertComparisonOp(n.Operator)
+			n.Left, n.Right = n.Right, n.Left
+		} else if !IsConst(n.Right) {
+			return v, n
+		}
+
+		left, ok := n.Left.(*BinaryExpr)
+		if !ok {
+			return v, n
+		}
+
+		// The right is const and the left side is a binary expression. Rotate the
+		// comparison combining portions that are const.
+
+		switch {
+		case IsConst(left.Right):
+			//        cmp          cmp
+			//       /   \        /   \
+			//    [+-/]   2  ->  a   [-+*]
+			//   /     \            /     \
+			//  a       1          2       1
+
+			switch left.Operator {
+			case Plus, Minus, Div:
+				n.Left = left.Left
+				left.Left = n.Right
+				if left.Operator == Plus {
+					left.Operator = Minus
+				} else if left.Operator == Minus {
+					left.Operator = Plus
+				} else {
+					left.Operator = Mult
+				}
+				n.Right, v.err = EvalExpr(left)
+				if v.err != nil {
+					return nil, nil
+				}
+				if !isVar(n.Left) {
+					// Continue as long as the left side of the comparison is not a
+					// variable.
+					continue
+				}
+
+				// TODO(pmattis): Handle Div?
+			}
+
+		case IsConst(left.Left):
+			//       cmp              cmp
+			//      /   \            /   \
+			//    [+-]   2  ->     [+-]   a
+			//   /    \           /    \
+			//  1      a         1      2
+
+			switch left.Operator {
+			case Plus, Minus:
+				left.Right, n.Right = n.Right, left.Right
+				if left.Operator == Plus {
+					left.Operator = Minus
+					left.Left, left.Right = left.Right, left.Left
+				} else {
+					n.Operator = invertComparisonOp(n.Operator)
+				}
+				n.Left, v.err = EvalExpr(left)
+				if v.err != nil {
+					return nil, nil
+				}
+				n.Left, n.Right = n.Right, n.Left
+				if !isVar(n.Left) {
+					// Continue as long as the left side of the comparison is not a
+					// variable.
+					continue
+				}
+
+				// TODO(pmattis): Handle Div?
+			}
+		}
+
+		// We've run out of work to do.
+		return v, n
+	}
+}
+
+func invertComparisonOp(op ComparisonOp) ComparisonOp {
+	switch op {
+	case EQ:
+		return EQ
+	case GE:
+		return LE
+	case GT:
+		return LT
+	case LE:
+		return GE
+	case LT:
+		return GT
+	default:
+		panic(fmt.Sprintf("unable to invert: %s", op))
+	}
+}
+
+type isConstVisitor struct {
+	isConst bool
+}
+
+var _ Visitor = &isConstVisitor{}
+
+func (v *isConstVisitor) Visit(expr Expr, pre bool) (Visitor, Expr) {
+	if pre && v.isConst {
+		switch expr.(type) {
+		case DReference, *ExistsExpr, *QualifiedName, *Subquery, ValArg:
+			v.isConst = false
+			return nil, expr
+		}
+	}
+	return v, expr
+}
+
+// IsConst returns true if the expression contains only constant values
+// (i.e. it does not contain a DReference).
+func IsConst(expr Expr) bool {
```

Comment:
It was used by index selection during the development of the PR, but that has since changed. I've unexported it.